### ISSUE TITLE
Add GeoPackage layer as input to model #30827 #38607

### ISSUE THREAD

**Issue by feefladder** [qgis/QGIS#45905]
### Feature description

Currently, it is not possible to add a GeoPackage layer (or `.xlsx` worksheet) as a model input, since the `Browse for Layer...` option is not available:
![image](https://user-images.githubusercontent.com/33122845/140554245-49cedbb9-451a-469b-a6db-2721b86214c0.png)

### Additional context

This is very important for a modeling workflow that uses GeoPackages. It *is* possible to save model outputs both inside models:
![image](https://user-images.githubusercontent.com/33122845/140554651-d56f1312-2299-41f8-b505-a0d5303d281d.png)
As well as when the model is run.
![image](https://user-images.githubusercontent.com/33122845/140554723-c14d2378-4c05-4b80-ae74-808ec9e94a5f.png)

But then loading them is difficult. There is this [workaround](https://gis.stackexchange.com/a/329363/156742), which is far from optimal and I cannot get it to work atm. Also manually entering `|layername=<whatever>` after browsing for the path should result in the same I think, but this does not work currently.

**Comment by gioman** [comment]
@Joeperdefloep I cannot confirm (tested on QGIS 3.22.0)

![image](https://user-images.githubusercontent.com/1951107/140560879-1d7537a7-a066-4b2b-9ca8-399886b72548.png)

anyway I think you are doing it wrong. If you want to hardcode an input in your model, you should do it within the tool that has to process that input (choosing manually the input and not using the "model input" option).

**Comment by Joeperdefloep** [comment]
The problem is not that you cannot specify a GeoPackage for a vector layer input, but that you cannot further specify the `layername` within the GeoPackage. So if I have a GeoPackage with two vector layers `foo` and `bar`, there is no way with the pop-up menu to select these.

However, in most tools where a Vector Layer can be selected, there is a drop-down with a *Browse for Layer...* option:
![image](https://user-images.githubusercontent.com/33122845/140592824-5ede433f-97f3-4421-a289-c21221257c8a.png)
where we can further specify which layer to select:
![image](https://user-images.githubusercontent.com/33122845/140592846-9311e3ea-6a9f-4c27-b37e-c349b58787f6.png)
But this is not possible when using a Raster Layer or Vector Layer input, because the *Browse for Layer...* option is not available.

**Comment by Joeperdefloep** [comment]
The problem seems to be [here](https://github.com/qgis/QGIS/blob/2d1aa68f0d044f2aced7ebeca8d2fa6b754ac970/src/gui/processing/qgsprocessingmaplayercombobox.cpp#L91): It only shows the popup for a `QgsProcessingParameterFeatureSource`, which in my understanding is a more broad form of a `QgsProcessingParameterVectorLayer`. I think changing this to:
```
mParameter->type() == QgsProcessingParameterFeatureSource::typeName() || mParameter->type() == QgsProcessingParameterVectorLayer::typeName()
```
should work. Also, #38607 could maybe be resolved by adding another `if else` statement testing for `QgsProcessingParameterRasterLayer` and refactoring [here](https://github.com/qgis/QGIS/blob/2d1aa68f0d044f2aced7ebeca8d2fa6b754ac970/src/gui/processing/qgsprocessingmaplayercombobox.cpp#L665-L669) to accept a `QgsMapLayerType::RasterLayer`. I'm currently compiling QGIS from source and may be able to submit a PR

**Comment by Joeperdefloep** [comment]
A current workaround is to convert the model to a script, and changing the `QgsProcessingParameterVectorLayer` to a `QgsProcessingParameterFeatureSource`. What is the difference between the two actually?

**Comment by nyalldawson** [comment]
Feature sources are for inputs which operate on the features in a layer (buffer, reproject,.etc) vs a vector layer input which operates on an entire layer at once (adding fields, removing fields)

**Comment by feefladder** [comment]
Closed by #45924
